### ISSUE TITLE
Setting up checkstyle with flycheck for java

### ISSUE THREAD

**Issue by SreenivasVRao** [flycheck/flycheck#1788]
This template is for **user support**.  If you are reporting an actual Flycheck bug, please use the relevant issue template instead.

## Checklist
- [ x] I have searched existing issues and StackOverflow for solutions.
- [ x] I have read the [Flycheck manual][].
- [x ] I have read the [Troubleshooting Guide][].
- [x ] I have checked my issue has not been reported yet.

## Problem description
Flycheck warns about “non-zero exit code, but no errors”

## Steps to reproduce
1. Open a java file.
2. Use custom checker
```
 (flycheck-define-checker my-java-checker
   " a java checker "
    :command ("java" "-jar /Users/venkobas/.config/emacs/java/checkstyle-8.32-all.jar" " -c /sun_checks.xml" source)
    :error-parser flycheck-parse-checkstyle
    :enable t
    :modes (java-mode))
```
3. See error
```
Suspicious state from syntax checker my-java-checker: Flycheck checker my-java-checker returned 1, but its output contained no errors: Unrecognized option: -jar /Users/venkobas/.config/emacs/java/checkstyle-8.32-all.jar
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

Try installing a more recent version of my-java-checker, and please open a bug report if the issue persists in the latest release.  Thanks!
```

- [ x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## System configuration
```
Syntax checkers for buffer Test.java in java-mode:

First checker to run:

  my-java-checker (explicitly selected)
    - may enable: yes
    - executable: Found at /usr/bin/java

Checkers that could run if selected:

  lsp  select
    - may enable: yes
    - may run:    t

The following syntax checkers are not registered:
  - my-java-checker
Try adding these syntax checkers to `flycheck-checkers'.

Flycheck Mode is enabled. Use M-c u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20200516.1719)
Emacs version:    26.3
System:           x86_64-apple-darwin18.2.0
Window system:    ns

```

Emacs configuration:
- [ ] Plain Emacs / Custom configuration
- [ x] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes

[Flycheck manual]:  https://www.flycheck.org/en/latest/user/troubleshooting.html
[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by SreenivasVRao** [comment]
Hi,

I'm able to run the same command in a terminal and see the error report:

```
java -jar ~/.config/emacs/java/checkstyle-8.32-all.jar -c /sun_checks.xml Test.java
```

I'm not sure why it fails to work with flycheck. I know java isn't supported. So any ideas/tips on what to do?

I'm able to get syntax highlighting via lsp checker. I'd like ideally, both lsp checker and this style checker to work. Can you tell me how I could achieve that?

Also, thanks a lot for such a great package!

**Comment by SreenivasVRao** [comment]
Looks like I  had some issues with writing the command. Seeing a slightly different error:

```
 (flycheck-define-checker my-java-checker
   " a java checker "
    :command ("java" "-jar" "/Users/venkobas/.config/emacs/java/checkstyle-8.32-all.jar" "-c" "/sun_checks.xml" source)
    :error-parser flycheck-parse-checkstyle
    :enable t
    :modes (java-mode))
```

```
Suspicious state from syntax checker my-java-checker: Flycheck checker my-java-checker returned 22, but its output contained no errors: Starting audit...
[ERROR] /var/folders/9v/40nj82cd4c1506g44nwstv5j7ch2p5/T/flycheckcU1WV0/MyClient.java:1: Missing package-info.java file. [JavadocPackage]
[ERROR] /var/folders/9v/40nj82cd4c1506g44nwstv5j7ch2p5/T/flycheckcU1WV0/MyClient.java:27:5: Missing a Javadoc comment. [JavadocVariable]
...
...
Audit done.
Checkstyle ends with 22 errors.
```


What should I be doing here..?

**Comment by cpitclaudel** [comment]
The output isn't xml, so the checkstyle XML parser is rejecting it.  There is likely a flag to pass to checkstyle to change the output?

**Comment by SreenivasVRao** [comment]
Hi, yes thank you for the quick reply.

```
 (flycheck-define-checker my-java-checker
   " a java checker "
    :command ("java" "-jar" "/Users/venkobas/.config/emacs/java/checkstyle-8.32-all.jar" "-c" "/sun_checks.xml" "-f" "xml" source)
    :error-parser flycheck-parse-checkstyle
    :enable t
    :modes (java-mode))
```

fixes it.

Can you tell me how I can have multiple checkers active? I'd like both lsp checker and this one to work.

**Comment by cpitclaudel** [comment]
See https://github.com/flycheck/flycheck/issues/1762 — we're working on it

**Comment by cpitclaudel** [comment]
In your case, flycheck-add-next-checker would work

**Comment by Dreameh** [comment]
I have been working to make checkstyle be a checker for flycheck, and it seems like the wonderous line of:
`Checkstyle ends with 22 errors.` 
which happens at the very end of each checkstyle run is what makes flycheck not like it while I was doing flycheck-compile of the checker.

There's a fix that can be made for checkstyle by removing the line itself, however, might there be a way to filter out that line so it will compile?

**Comment by cpitclaudel** [comment]
> which happens at the very end of each checkstyle run is what makes flycheck not like it while I was doing flycheck-compile of the checker.

Can you clarify what the issue is?

**Comment by Dreameh** [comment]
The issue at hand is this:
```xml
-*- mode: compilation; default-directory: "~/Projects/emacs-testing/src/main/java/dev/dreameh/" -*-
Compilation started at Mon Nov 14 04:47:56

"c:/Program Files/Eclipse Adoptium/jdk-17.0.4.101-hotspot/bin/java.exe" "-jar" "c:/Users/kyusa/lang-servers/checkstyle-8.14-all.jar" "-c" "c:/Users/kyusa/lang-servers/docs/checkstyle.xml" "-f" "xml" "c:/Users/kyusa/Projects/emacs-testing/src/main/java/dev/dreameh/GreetingResource.java"
<?xml version="1.0" encoding="UTF-8"?>
<checkstyle version="8.14">
<file name="c:\Users\kyusa\Projects\emacs-testing\src\main\java\dev\dreameh\GreetingResource.java">
<error line="1" severity="error" message="File does not end with a newline." source="com.puppycrawl.tools.checkstyle.checks.NewlineAtEndOfFileCheck"/>
<error line="1" severity="error" message="Missing package-info.java file." source="com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocPackageCheck"/>
<error line="8" severity="error" message="Missing a Javadoc comment." source="com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocTypeCheck"/>
<error line="11" column="3" severity="error" message="Missing a Javadoc comment." source="com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocVariableCheck"/>
<error line="11" column="31" severity="error" message="Name &apos;test_file&apos; must match pattern &apos;^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$&apos;." source="com.puppycrawl.tools.checkstyle.checks.naming.ConstantNameCheck"/>
<error line="13" column="3" severity="error" message="Class &apos;GreetingResource&apos; looks like designed for extension (can be subclassed), but the method &apos;hello&apos; does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class &apos;GreetingResource&apos; final or making the method &apos;hello&apos; static/final/abstract/empty, or adding allowed annotation for the method." source="com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheck"/>
<error line="13" column="3" severity="error" message="Missing a Javadoc comment." source="com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocMethodCheck"/>
</file>
</checkstyle>
Checkstyle ends with 7 errors.

Compilation exited abnormally with code 7 at Mon Nov 14 04:47:57
```
As one can see, the line "Checkstyle ends with 7 errors" directly influences the compilation thinking it exits abnormally with code 7, which is where my question comes in, if there's a way to just filter away the Checkstyle ends with X errors as that line is only saying how many errors/warnings/info there were in the code itself.

**Comment by cpitclaudel** [comment]
> As one can see, the line "Checkstyle ends with 7 errors" directly influences the compilation thinking it exits abnormally with code 7

But why is that an issue?  Compilation exited abnormally is just what M-x compile prints when the compiler returns a non-zero exit code, which is a good thing here.  If you filter the output that message won't change (it's just the processes' error code).  Try `M-x compile RET exit 7 RET`:

```
-*- mode: compilation; default-directory: "~/.emacs.d/" -*-
Compilation started at Sun Nov 13 21:51:22

exit 7

Compilation exited abnormally with code 7 at Sun Nov 13 21:51:22
```

But yes, it should be easy to filter it by changing the error parser to remove that line.
